### ISSUE TITLE
Cloud service - false blockings ?

### ISSUE THREAD

**Issue by mikeruss1** [CIDRAM/CIDRAM#76]
the vast majority of my recaptcha passes are the result of a Cloud service block. As far as I can tell the user is not a bot, generally he passes the recaptcha and views the blocked page and maybe 1 or 2 others, then moves on. No evidence of a bot. This happens once or twice a day.

I dont know what is causing this - maybe VPNs ? Can we do anything ?

**Comment by Maikuolan** [comment]
> maybe VPNs ?

Could be. Could also quite possibly still be a "cloud service" (thus, a correct inference for the description of the reason blocked), while at the same time, also being a human (thus, a false positive, in the sense that when we block something for the reason of being a cloud service, it tends to imply that connections from such addresses aren't ever going to originate from a human).

This is because the old dichotomy of a connection being either a human endpoint or a cloud endpoint doesn't necessarily always hold true anymore, in part due to the rise and increasing prevalence nowadays of modern "multi-play networks" ("triple play", "quad play" networks, etc).

A very quick explanation: A "single-play" network refers to a network that tends to do just one thing only, whether that's web hosting, or domestic internet connectivity, or business infrastructure, or whatever else.  When networks start doing more than one thing (e.g., a web hosting provider one day decides that they can make some more money by also providing domestic internet connectivity in the local area where their servers reside, thus meaning that from that point onward, connections from their network could be from either a webhosting account, or from a human endpoint; it might or mightn't be so easy to tell which is which, depending on how they design their network, announce specific prefixes and so on), then they could be considered as a "dual-play", "triple-play", "quad-play" network and so forth, depending on how many different services they provide. A typical "quad-play" network, for example, might provide webhosting, digital TV services, domestic internet connectivity, and business infrastructure and/or co-location ("colo") services.

This makes our job a little more difficult unfortunately, but is predominantly our problem though, rather than theirs, and is a trend likely to continue into the future, I think. Usually, to play it safe, if I discover that a network listed in the signature files as a "cloud service" is actually a multi-play network, and if a significant proportion of requests seen from that network are from things that aren't related to cloud services, I'll reassess whether it's worth continuing to block them, and if I determine that it's worth it, I'll shift them out from the cloud signature files and into the ISPs signature files and change the block reason to "generic" (which would just mean we'd be blocking them, without implying anything about why we're blocking them by the cited block reason specifically, and I may possibly also mark them for use with reCAPTCHA if it seems worthwhile doing so), or if it's not worth it anymore, just remove them from the signature files entirely.

There are also some networks listed in the cloud signature files which are technically multi-play, providing both webhosting services and VPN services, whereby I've kept them there, but I've been able to mark specific prefixes distinctly as being proxies or VPNs, rather than cloud services, to keep the cited block reasons accurate (this in cases where prefix descriptions have been more detailed, or properly descriptive, allowing me to better figure out what's what, or cases where information is otherwise available somewhere, like from the networks themselves, to allow me to figure out what's what).

Do you have copies of any log entries corresponding to these? The best way to proceed would be the reassess networks associated with the signatures in question to determine whether anything needs to be changed in the signature files, but to do this, I'll need to know which IPs were blocked.

**Comment by mikeruss1** [comment]
I have them all - give me a while pl

**Comment by mikeruss1** [comment]
a few recent examples ...
IP Address: 103.86.96.14 

IP Address: 199.241.120.78 

IP Address: 86.123.49.178 

IP Address: 38.121.94.102 

IP Address: 104.223.95.2 

IP Address: 109.169.61.31 

IP Address: 103.238.131.205 

IP Address: 172.105.205.236 

IP Address: 142.4.124.196 

IP Address: 159.65.28.193 

IP Address: 86.121.187.70

**Comment by mikeruss1** [comment]
and just got this one ...

IP Address: 83.241.25.42

**Comment by mikeruss1** [comment]
and one from today ...
IP Address: 5.62.43.14

**Comment by Maikuolan** [comment]
38.121.94.102 is a false positive, caused by a bug in the default signature bypasses. It should be being reported as "generic", rather than "cloud service" (i.e., it's intended to be blocked by the signature files, but the cited reason for it is wrong). Working on fix right now. Still investigating the others.

**Comment by Maikuolan** [comment]
*38.121.94.102 false positive fixed via 3568ab5ea136f0fb8a1d7b7a412a12e4425ffd20.*

**Comment by Maikuolan** [comment]
86.121.187.70 and 86.123.49.178 belong to "RCS & RDS SA". Numerous active spam events from their ASN and good reasons to block them, but still false positives, due to that their network seems to be quad-play, so should be listed as "generic" rather than "cloud service". Will fix that with the next signatures update.

*Edit: Done.*

**Comment by Maikuolan** [comment]
103.86.96.14 is now listed as a proxy, rather than a cloud service. Still working through the others.

**Comment by mikeruss1** [comment]
one more ...
 IP Address: 172.105.209.138

**Comment by mikeruss1** [comment]
and one more, I wont send anymore unless you request ...
5.62.39.227

**Comment by Maikuolan** [comment]
No problem either way. I'll get through them all eventually anyhow. Just haven't put a deadline on it due to an unpredictable schedule recently.

**Comment by mikeruss1** [comment]
I guess the ideal would be if real users using generic or proxy (if thats the right terms) were let through automatically without having to go through recaptcha. It seems this isnt possible ?

**Comment by Maikuolan** [comment]
The main problem is, how do we *know* who the "real" users are. If they're not connecting via normal domestic endpoints, but instead, through endpoints already known to be regularly used and exploited by spammers, spambots, scrapers, hackers, and etc, how do we know that they are any different? One way, is by asking them to complete the reCAPTCHA instance (some bots are capable of completing CAPTCHAs, including reCAPTCHA, in the same way as a normal human user would do, but most of them aren't, or just don't bother trying; most human users will at least try to complete a CAPTCHA if presented with one, and should pass in most cases, if it isn't impossible to read, understand, or use properly). Especially when most bots will do everything in their power to *appear* as human as possible (e.g., by faking normal browser UAs, attempting to access requests in a reasonably human-looking manner, etc).

The process can be simplified by using the "invisible" API (whereby JavaScript is leveraged to make their browser complete reCAPTCHA for them automatically, so that they don't need to click on anything or do any extra work in most cases), though that may be problematic for websites serving primarily EU users (due to cookie laws and such). Should okay for elsewhere though.

Of course, if we wanted to just allow *all* generic and/or proxy requests through, that can be achieved by configuring the package accordingly (i.e., disabling the `block_generic` and/or `block_proxies` directives accordingly), and that might be the best bet, if we wanted to be as confident as possible that humans aren't being blocked (although it may also allow unwanted humans in too, which might try to do unwanted things on websites manually, like posting spam and such; this is also the main principle behind blocking proxies, and identifying proxies as such).

Actually identifying whether a request originates from a human or a bot though, can never be fully 100% guaranteed, and is a job which becomes more difficult, when human users start trying to connect to websites via networks that don't typically provide human endpoints or direct internet connectivity otherwise. All we have, after all, is the request, and any details pertinent to it, like IP address, user agent, etc. There are also cases where some networks are listed, not because of bots, but because ~99% of the traffic seen from them are from human spammers, hackers, and that type of thing.

Always open for new ideas though.

**Comment by Maikuolan** [comment]
I've finished going through all the rest. Couldn't find any evidence at all of them being anything other than cloud services (webhosting providers, server racks, etc). That, of course, doesn't necessarily mean that the owner of a server or website, couldn't simply choose to run a proxy script or similar such thing to use their own webspace or server rack as a private proxy for themselves, or that there are some hosting services there running proxy services on the side in private. However, there's nothing publicly announced anywhere for the mentioned IPs AFAICT, nothing announced as proxies, VPNs, human endpoints or etc for them, so that, with them announcing themselves as cloud services of some form or another, and no evidence otherwise of them being anything other than cloud services, I think labeling them as such, at least for now, remains the most accurate label we could use for what to cite when blocking them.

I did remove a few cloud services and related networks from the signature files over the past few days that I didn't think needed to be blocked anymore anyway (no recent evidence of bad behaviour like spamming or unsolicited bot requests coming from their networks and such), so the signature files did very briefly become thinner than usual (..for the 24-hours or so that it's worth, until additions elsewhere brought them back to their usual density again, anyway.. I sense possible manageability problems in the future in that regard, but that's a separate issue and separate topic of discussion anyway), but none of those removed networks included the aforementioned IPs.

So, in summary of all work done: Some false positives confirmed and consequently either removed or relabeled, though most of the aforementioned IPs have remained as they were.

Marking as *fixed* and closing.
